### ISSUE TITLE
Download pictures to local

### ISSUE THREAD

**Issue by jinsen711** [obsidianmd/obsidian-clipper#37]
Can the images in the web page be automatically downloaded when saving the web page? I am using the "Custom Attachment Location" plug-in to manage the image files in each markdown in obsidian.

**Comment by kepano** [comment]
Not yet, but there is a plugin called Local Images that can do this for you.

**Comment by jinsen711** [comment]
> Not yet, but there is a plugin called Local Images that can do this for you.

Thanks

**Comment by marconoris** [comment]
> Can the images in the web page be automatically downloaded when saving the web page? I am using the "Custom Attachment Location" plug-in to manage the image files in each markdown in obsidian.

+1!

**Comment by codey-badger** [comment]
> > Not yet, but there is a plugin called Local Images that can do this for you.
> 
> Thanks

Hi, is it working automatically or you have to confirm each time?

**Comment by LeCheenaX** [comment]
+1! Great to have the option to determine if using local images or online images.

**Comment by jinsen711** [comment]
> > > Not yet, but there is a plugin called Local Images that can do this for you.
> > 
> > 
> > Thanks
> 
> Hi, is it working automatically or you have to confirm each time?

Custom Attachment Location plugin? It's automatic. Sorry I just saw it now.

**Comment by Tback1** [comment]
> Not yet, but there is a plugin called Local Images that can do this for you.

It not work everywherel. Because some image you could see on the brower which has full refrence even cookie from page, the CDN will not block the request. but when you capture it to the Obsidian ,it will not.
So may be need get the image from the Brower's cache or download from Brower

**Comment by emptizen** [comment]
May I ask if there is any plan to make it a new feature?

**Comment by kepano** [comment]
Yes but it will require changes to Obsidian URI, so it will take some time.

**Comment by guitarbug** [comment]
> Not yet, but there is a plugin called Local Images that can do this for you.

I have the plugin "Local Images Plus" installed, When I use Web Clipper to save the web page with pictures, the pictures are not download to local.
but I use "MarkDownload - Markdown Web Clipper" to save the same page, the pictures are download to local.

**Comment by myInstagramAlternative** [comment]
I just tested base64-encoded images and Obsidian renders them.

Would it be possible to add this as an option? I'm not sure how difficult it is to extract images from the page, but inserting the image in base64 format into Obsidian should be pretty easy to implement.

`![imagel](data:image/png;base64,encoded_image_string)`

**Comment by kepano** [comment]
Storing the images as base64 is not a good solution. It will lead to Markdown files that are multiple megabytes which hinders the performance of Markdown parsing and will cause syncing issues.

**Comment by Creveoolus** [comment]
> Not yet, but there is a plugin called Local Images that can do this for you.

So I just downloaded obsidian web clipper but I missing this function aswell.
My problem is that Local Images plugin cant download pictures if the site requires authorization, but web clipper could.

> Yes but it will require changes to Obsidian URI, so it will take some time.

Is there an ETA for it?

Thank you!

**Comment by sourcecodemage** [comment]
+1 for this feature. I just tested the plugin and , after appearance, attachment location was the second thing I checked. I download a lot of wiki pages with instructions and screenshots of what the app or screen output should look like mixed in with the instructions, so  if the image link ever breaks , then I'll have a lot of blank areas in my note.   

e.g. 

`5. Type this command. If the output doesn't look like the below screenshot, then return to previous step`.

Otherwise I really like the plugin. I still need to figure out how to tell it which vault images should go into, but I only just installed it and have read zero documentation. The attachments concern sidetracked me from that!

**Comment by greetclammy** [comment]
+1.

Local images are a must. For privacy, offline use and future-proofing.

Feature request for the *Local images plus* plugin to auto download clipped content: https://github.com/Sergei-Korneev/obsidian-local-images-plus/issues/87#issuecomment-2523915846

**Comment by kepano** [comment]
This is now possible using Obsidian 1.8.0 (currently in beta)
https://obsidian.md/changelog/2024-12-18-desktop-v1.8.0/

**Comment by chenxiccc** [comment]
> This is now possible using Obsidian 1.8.0 (currently in beta) https://obsidian.md/changelog/2024-12-18-desktop-v1.8.0/

Sorry, I don't understand. 
Do you mean that:
1. Clip the webpage by Obsidian web clipper
2. In obsidian, press the hotkey for Download attachments for current file
Will you consider to add a setting, such as 
- [ ] Download attachments
 in the web clipper and make it just one step to download all img/attachements of the clipped webpage?

**Comment by Tback1** [comment]
> > This is now possible using Obsidian 1.8.0 (currently in beta) https://obsidian.md/changelog/2024-12-18-desktop-v1.8.0/
> 
> Sorry, I don't understand. Do you mean that:
> 
> 1. Clip the webpage by Obsidian web clipper
> 2. In obsidian, press the hotkey for Download attachments for current file
>    Will you consider to add a setting, such as
> 
> * [ ]  Download attachments
>   in the web clipper and make it just one step to download all img/attachements of the clipped webpage?

it's not solve the question that the url need ref from origion source.espically some CDN

**Comment by chenxiccc** [comment]
I've created a plugin to solve this.
https://github.com/chenxiccc/obsidian-auto-download-imgs-after-web-clipping

**Comment by rianbk** [comment]
@Tback1 has it right I think. **Download attachments for current file** runs inside Obsidian, so it doesn't have the browser's cookies or the page's referrer. That's fine for public images, but anything behind a login just fails, and you sometimes end up with a login page saved as a `.png`.

The only time you can get those images is while you're still signed in, in the browser. So I had a go at doing it in the extension. It's opt-in and off by default:

https://github.com/rianbk/obsidian-clipper/tree/save-images-when-clipping

@kepano, you mentioned this would need Obsidian URI changes. This way doesn't. The note is handed over exactly as it is now, and images are fetched and saved separately.

It tries three ways to fetch each image, because no single one covers everything. The content script has the page's cookies but is limited by the page's CORS rules. The background worker isn't, but doesn't send the same cookies. The downloads API has all the cookies but no referrer. If all three fail, that image just keeps its original link so the note still saves fine.

A few things I got wrong at first, in case they're useful:

- Filenames need deciding before any downloading starts. Otherwise the order they finish in decides the names, and re-clipping a page can give the same name to a different image.
- You have to check the file's actual bytes to know its type. One site I tested says `image/jpeg` while sending a PNG.
- You have to check the response is really an image. Links like `/wiki/File:X.jpg` are web pages, and they return 200, so Wikipedia's front page gave me five HTML files named `.jpg` before I noticed.

I've tested it on pages behind a login, on Wikipedia, and after browser restarts. Safari has no downloads API so the setting just hides itself there.

This is a big change that nobody asked me for, so I thought I'd ask before opening a PR. The branch also has optional support for the Local REST API plugin, which I'd take out for anything upstream, since relying on a community plugin feels like your call.

Very happy to split it up, change the approach, or just leave it as a fork if you'd rather do it differently.
